org-code-javabuilder/lib/src/main/java/org/code/javabuilder/AWSContentManager.java

```diff
@@ -0,0 +1,183 @@
+      throw new UserInitiatedException(UserInitiatedExceptionKey.TOO_MANY_UPLOADS);
+    }
+    final String key = this.generateKey(filename);
+    final long expirationTimeMs = System.currentTimeMillis() + context.getRemainingTimeInMillis();
+
+    try {
+      final URL presignedUrl =
+          s3Client.generatePresignedUrl(
+              this.bucketName, key, new Date(expirationTimeMs), HttpMethod.PUT);
+      this.uploads++;
+      // Add the GET url for this file to the asset map so it can be referenced later.
+      this.projectData.addNewAssetUrl(filename, this.getContentUrl + "/" + key);
+      return this.getContentUrl + presignedUrl.getFile();
+    } catch (AbortedException e) {
+      // this is most likely because the end user interrupted program execution. We can safely
+      // ignore this.
+    } catch (SdkClientException e) {
+      throw new InternalServerError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e);
+    }
+
+    return null;
+  }
+
+  @Override
+  public String writeToOutputFile(String filename, byte[] inputBytes, String contentType)
+      throws JavabuilderException {
+    if (this.writes >= WRITES_PER_SESSION) {
+      throw new UserInitiatedException(UserInitiatedExceptionKey.TOO_MANY_WRITES);
+    }
+    String filePath = this.generateKey(filename);
+    ObjectMetadata metadata = new ObjectMetadata();
+    metadata.setContentType(contentType);
+    metadata.setContentLength(inputBytes.length);
+    ByteArrayInputStream inputStream = new ByteArrayInputStream(inputBytes);
+
+    try {
+      this.s3Client.putObject(this.bucketName, filePath, inputStream, metadata);
+    } catch (AbortedException e) {
+      // this is most likely because the end user interrupted program execution. We can safely
+      // ignore this.
+    } catch (SdkClientException e) {
+      // We couldn't write to S3, send a message to the user and fail. The S3 SDK includes retries.
+      throw new InternalServerError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e);
+    }
+
+    this.writes++;
+    return this.getContentUrl + "/" + filePath;
+  }
+
+  @Override
+  public void verifyAssetFilename(String filename) throws FileNotFoundException {
+    try {
+      // We should only hit this exception if we try to verify an asset before source code has
+      // been loaded, which should only be in the the case of manual testing. Log this exception but
+      // convert to a FileNotFoundException to preserve the method contract.
+      // Note / TODO: Once we fully migrate away from Dashboard sources, we can remove the
+      // loadProjectDataIfNeeded() call here and this exception handling.
+      this.loadProjectDataIfNeeded();
+    } catch (InternalServerError e) {
+      throw new FileNotFoundException("Error loading data");
+    }
+    if (!this.projectData.doesAssetUrlExist(filename)) {
+      throw new FileNotFoundException(filename);
+    }
+  }
+
+  private String generateKey(String filename) {
```

Comment:
Can you add a comment here? I think I know what key it's generating, but I'm pretty sure me in a few weeks will have forgotten enough context to get confused :)